src/HttpHeader.cc

```diff
@@ -486,6 +488,15 @@ HttpHeader::parse(const char *header_start, size_t hdrLen, Http::ContentLengthIn
             return 0;
         }
 
+        const bool frameHeader = (e->id == Http::HdrType::CONTENT_LENGTH || e->id == Http::HdrType::TRANSFER_ENCODING);
+        if (seenLines > 1 && frameHeader) {
+            debugs(55, warnOnError, "WARNING: obs-fold seen in " << e->name << ": " << e->value);
+            delete e;
+            PROF_stop(HttpHeaderParse);
+            clean();
+            return 0;
```

Comment:
We should not return 0 here -- the header is parseable -- but it is difficult to fix this properly because this PR was split from PR #702.  Assuming you insist on keeping the two PRs separate, I will insist on this PR being merged first with an XXX to be properly addressed in PR #702 (where the next generation of the framing field interpreter class can handle this nicely):

```suggestion
            return 0; // XXX: Signal invalid framing instead of unparsable header
```

Reply:
The protocol violations introduced to non-HTTP/1.1 protocols by banning obs-fold entirely instead of taking the "or replace" criteria are not related to and cannot be fixed by the additions in PR 702.

Reply:
Sorry, I have no idea what protocol violations you are talking about, and why any of this matters for this change request.

Reply:
Returning "Invalid Request" error in protocols where folding is a valid mechanism is a violation of those protocols. There are currently 3 non-HTTP/1.1 protocols using this parser where that will happen. Resolving this conflict of requirements will require the next level of ParserNG to be implemented - which is far too big a change to make a prerequisite for this and/or PR 702 (given that they need to backport).

Reply:
Returning "Invalid Request" error in protocols where folding is a valid mechanism is a violation of those protocols. There are currently 3 non-HTTP/1.1 protocols using this parser where that will happen.

Resolving this conflict of requirements will require the next level of ParserNG to be implemented - which is far too big a change to make a prerequisite for this and/or PR 702 (given that the need to backport).

Reply:
> Returning "Invalid Request" error in protocols where folding is a valid mechanism is a violation of those protocols. There are currently 3 non-HTTP/1.1 protocols using this parser where that will happen.

I do not see how your comment applies to my change request. I am asking to add an XXX for the new code. You are responding with a statement that implies that the new code violates some protocols. If you are right, that is a yet another reason to add an XXX! 

> Resolving this conflict of requirements will require the next level of ParserNG to be implemented - which is far too big a change to make a prerequisite for this and/or PR 702 (given that they need to backport).

Let's argue about that in PR 702, after it gets this XXX from this PR. It is very difficult to have a reasonable conversation about the future of the code in a different PR while discussing a specific ("add an XXX") change request in another PR. The XXX itself is valid and should be added. I suggest postponing other discussions until we have only one PR (with a lot more relevant-to-the-discussion code) to deal with.

Please do not misinterpret this comment as an agreement with your statements. Fortunately, no such agreement is necessary to add that XXX.

Reply:
Adding the XXX. But FYI PR 702 does not contain the next generation of Parser - all that PR has is another layer of partial improvement (hacks) to cope with current problems (targeted at v5 and v4 usage). The actual next-gen proposal is PR 703 (targeted at v6).

Reply:
XXX Done in 9d4d1f555. If I still misunderstood what you are requesting please suggest specific wording.

Reply:
> But FYI PR 702 does not contain the next generation of Parser

Yes, I know. I believe that all we should do to address the underlying problem that spawned PR 701 and PR 702 (so far) can be done without upgrading/rewriting a lot of parsing code.

> targeted at v5 and v4 usage

We should fix the problem in master/v6 and then discuss what to do in v4 and what to do in v5. I certainly agree that v5 and/or v4 fixes may be different, but we should always start from the top -- that is how we have agreed to modify Squid.

If you want to treat this particular case as an exception, I can (reluctantly) agree to that as well, but we should postpone the current set of PRs and do a PR against v4 in that case. You would be able to commit virtually anything into v4, even if I think that more should be done in master/v6. Once that v4 PR is merged, we will pretty much ignore v4 when resuming our master/v6 work (instead of arguing whether every change is appropriate for master because of vague v4 porting concerns).

The decision about v5 should be done once master/v6 changes are known. If master/v6 changes are small/safe/surgical enough, then they should be merged into v5 as is IMO, because v5 is still very unstable right now -- there is no reason to create this significant difference for the sake of v5 stability -- and because future parsing work in the related areas would significantly benefit from this v5 exposure. Otherwise, v5 can get v4 changes (or even something in-between).

> If I still misunderstood what you are requesting please suggest specific wording.

This XXX is not related to the method API. It is related to a specific code snippet added by this PR. I _have_ [suggested](https://github.com/squid-cache/squid/pull/701#discussion_r463853974) specific XXX location and wording already!

Reply:
I see. The problem is that these cases are *not* invalid framing. They are perfectly valid syntax - just *suspicious* due to being unnecessary existing. It is a somewhat arbitrary policy decision to reject these messages instead of continuing with the status-quo ignore behaviour.

The same new mechanism to deliver signals to callers should be usable for all the other non-error situations this method handles. Thus my choice of placement and wording for the XXX. I intend to resolve that in PR 703 for v6 release before it moves out of WIP/Draft.

Reply:
> The problem is that these cases are not invalid framing. They are perfectly valid syntax... It is a somewhat arbitrary policy decision to reject these messages instead of continuing with the status-quo ignore behaviour.

I agree with the quoted summary, but I do not see "the problem" with that summary.

We should make a policy decision and then, if necessary, adjust the code to follow it.

* If our decision is to ban folded framing headers, then we should adjust the official code accordingly, eventually merging this PR.
* If the decision is to continue to allow folded framing headers, then this PR should be closed without merging.

My recommendation is (still) the former, but I do not insist on it. We can ban folded framing headers after the next CVE that will abuse folded framing headers (instead of speculating now that there will be such a CVE in the future).

The same logic applies to other things this PR bans in framing headers (currently "suspicious CRs" IIRC).